tableauserverclient/server/endpoint/datasources_endpoint.py

```diff
@@ -3,38 +3,48 @@
 from .permissions_endpoint import _PermissionsEndpoint
 from .fileuploads_endpoint import Fileuploads
 from .resource_tagger import _ResourceTagger
-from .. import RequestFactory, DatasourceItem, PaginationItem, ConnectionItem
+from .. import RequestFactory, DatasourceItem, PaginationItem, ConnectionItem, RequestOptions
 from ..query import QuerySet
 from ...filesys_helpers import to_filename, make_download_path, get_file_type, get_file_object_size
 from ...models.job_item import JobItem
+from ...models import ConnectionCredentials
 
 import os
 import logging
 import copy
 import cgi
 from contextlib import closing
 
+from pathlib import Path
+from typing import List, Dict, Sequence, Mapping, Optional, Tuple, Union, TYPE_CHECKING, Any, BinaryIO
+
 # The maximum size of a file that can be published in a single request is 64MB
 FILESIZE_LIMIT = 1024 * 1024 * 64   # 64MB
 
 ALLOWED_FILE_EXTENSIONS = ['tds', 'tdsx', 'tde', 'hyper']
 
 logger = logging.getLogger('tableau.endpoint.datasources')
 
+if TYPE_CHECKING:
+    from ..server import Server
+    from ...models import PermissionsRule
+
 
 class Datasources(QuerysetEndpoint):
-    def __init__(self, parent_srv):
+    def __init__(self, parent_srv: 'Server') -> None:
```

Comment:
I use a forward reference here because actually importing Server creates a circular dependency.  https://stackoverflow.com/questions/33837918/type-hints-solve-circular-dependency